[RUNTIME] Distributed Save and Load for Model

PKUFlyingPig opened the pull request:
I added two top-level APIs for Alpa: 
- save_checkpoint(ckpt_dir, target, step): the `target` is usually the trainState. This function (1) replaces the leaves of trainState (DistributedArray or ReplicatedDistributedArray) with the path to which the array is saved and saves this metadata for restore. (2) saves one replica for all the DistributedArrays/ReplicatedDistributedArrays in `target`.
- restore_checkpoint(executable, ckpt_dir, target, step): This function requires the DecentralizedRuntime which contains all the shardingSpec and mesh allocation info to load and reshard the DistributedArrays/ReplicatedDistributedArrays. 

Current version contains a defect that the trainState must be the first argument in the parallelized function (i.e. train_step). One possible fix requires the user to explicitly provide the index of the trainState in train_step's argument list. Any suggestion for a more elegant implementation?

zhisbug commented on alpa/__init__.py:
```diff
@@ -3,14 +3,15 @@
                               LocalPhysicalDeviceMesh,
                               DistributedPhysicalDeviceMesh, DistributedArray,
-                              fetch)
+                              ReplicatedDistributedArray, fetch)
```
do you use `ReplicatedDistributedArray` somewhere?

PKUFlyingPig replied:
I use it in the `serialization.py`.

PKUFlyingPig commented:
The root cause of the problem can be understood better with an example:

user's view: train_step(batch, train_state) where batch = {"x": x, "y": y}, train_state = {"a": a,  "params":{"w": w, "b": b}}, 
after tree_flatten: args = [x, y, a, w, b]
after split (batchnum = 2) : args = [x0, x1, y0, y1, a, w, b]

The shardingSpec and device mesh allocation information for all the arguments is stored according to the flatten and split version above. So in order to load train_state, I have to know the index of the first arg of train_state (a in this case) in the flatten and split argument arrays. 

After the discussion with @zhuohan123, we have two possible fixes:
(1) requires the user to explicitly provide the index of the trainState in train_step's argument list.
For the example above, the user provides index 1 in `restore_checkpoint`.
(2) @zhuohan123 argues that the user does not have to provide the entire `DecentralizedDistributedRuntime` into restore_checkpoint but only provide the shardingSpec and device mesh allocation info. To support this, we can add another API in `DecentralizedDistributedRuntime` to get the info.
For the example above, the user calls executable.get_load_info() to get the info list:
[{"x": x_info, "y": y_info}, {"a": a_info, "params": {"w": w_info, "b": b_info}}]
Then users can index into the list and get just enough info for loading and provide it into restore_checkpoint.

merrymercy commented:
This problem is very important. Let use move our discussion to https://github.com/alpa-projects/alpa/issues/439

zhuohan123 commented on alpa/serialization.py:
```diff
@@ -0,0 +1,149 @@
+import jax
+import msgpack
+import numpy as np
+from tensorflow.io import gfile
```
Why do we need this dependency?

PKUFlyingPig replied:
flax's serialization function uses these packages, maybe I can find some in-built utils to substitute?

zhuohan123 replied:
Where exactly does flax use this? I can't find it on this page:https://flax.readthedocs.io/en/latest/_modules/flax/serialization.html

PKUFlyingPig replied:
https://flax.readthedocs.io/en/latest/_modules/flax/training/checkpoints.html#save_checkpoint

zhuohan123 replied:
Let's leave this dependency here for now. However, we should remove the dependency to tensorflow just for this small part of function.

zhuohan123 commented on alpa/serialization.py:
```diff
@@ -0,0 +1,149 @@
+    replicated_distarray = 5
+
+
+def _ndarray_to_bytes(arr) -> bytes:
```
Can we just import this function from flax instead of copying it?

zhuohan123 commented on tests/test_dist_save_load.py:
```diff
@@ -73,20 +89,141 @@ def test_distributed_array_save_load(self):
+        assert_allclose(serial_state.params, parallel_state.params, 1e-3, 1e-3)
+
+        with tempfile.TemporaryDirectory(
+                prefix="/home/ubuntu/efs/") as ckpt_dir:
```
We should not include any absolute path. Can we somehow use ray to hack a temporary NFS just for testing purposes?

zhuohan123 replied:
Or you can force all the testing processes on a same node and just use the filesystem on the node.

zhisbug commented:
How's this issue going? will it get in main soon?

zhuohan123 commented:
From what I know this PR should be ready to merge now. @PKUFlyingPig Anything else you wanna do before we do a final round of review and merge the code?

PKUFlyingPig commented:
> From what I know this PR should be ready to merge now. @PKUFlyingPig Anything else you wanna do before we do a final round of review and merge the code?

This code requires tensorflow as a third-party package (flax's serialization needs this), should we mention it in the installation document and add it into the CI/DI scripts. This code failed the CI/Unittest due to this.

zhuohan123 commented on alpa/serialization.py:
```diff
@@ -0,0 +1,166 @@
+            load_info: shardingSpec and deviceMesh allocation info for loading.
+    """
+    ckpt_path = os.path.join(ckpt_dir, f"checkpoint_{step}")
+    with gfile.GFile(ckpt_path, 'rb') as fp:
```
Actually, if the only dependency to gfile is here, can we try to replace this gfile with python's native file and see whether the code still work for our cases?

PKUFlyingPig replied:
Sure, I will try.

merrymercy commented on alpa/pipeline_parallel/decentralized_distributed_runtime.py:
```diff
@@ -200,10 +203,17 @@ def __init__(self,
 
+        # Cached sharding indices for inputs.
+        # List[mesh_idx -> List[sharding_indices]].
+        self.input_indices = [] 
```
This var is deprecated

merrymercy reviewed (commented):
Let us remove the tensorflow dependency. Other things look good to me
